# Ebay Used Car Data Analysis

In this project we will be performing some EDA on a dataset involving cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. We'll start by cleaning our data and attempting to gain any insights the end results. 

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import re
#reading csv file
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Upon inspection, we observe that there is some data missing for a couple of columns in our data set. We also notice that our column names are in camelcase, which we will change to Python's preferred snakecase. In addition, we will also be rewriting a few of the column names

In [3]:
#rewriting column names
columns = autos.columns

new_columns = []
for c in columns:
    c = c.replace('yearOfRegistration', 'registration_year')
    c = c.replace('monthOfRegistration', 'registration_month')
    c = c.replace('noRepairedDamage', 'unrepaired_damage')
    c = c.replace('dateCreated', 'ad_created')
    c = re.sub( '(?<!^)(?=[A-Z])', '_', c ).lower()
    new_columns.append(c)
    
autos.columns = new_columns
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

From doing some basic data exploration, we can see a number of things:
    1. The 'price' and 'offer_type' don't have much variety in the data, so we can remove those columns.
    2. The data for the 'price' and 'odometer' columns are stored in text data rather than numerical.
    3. The max and min for 'registration_year' don't make any sense, so there might be incorrect data.
    4. There a big discrepancy between the max and the 75% values for the 'power_p_s' column, so there might be missing data.
    
The first thing we will do it convert the data in 'price' and 'odometer' to numerical data.

In [5]:
#converting price to numerical
autos['price'] = (autos['price']
                  .str.replace('$', '')
                  .str.replace(',','')
                  .astype(int)
                 )
#converting odometer to numerical
autos['odometer'] = (autos['odometer']
                    .str.replace('km', '')
                    .str.replace(',', '')
                    .astype(int)
                    )
#renaming odometer
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

The next step we will be taking is further analyzing data in our price and odometer_kg columns. We will take and quick look at several statistics and try to make a few notes on what we notice.

In [6]:
# Exploring price column
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().sort_index(ascending=False).head(30))
print(autos['price'].value_counts().sort_index().head(30))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price,

The first thing we notice about the price column is that there is a giant discrepancy between the max and 75% value from the describe function. Thus, there maybe some outliers (or data completely missing). We then notice that there is giant leap in values after 350,00 so we will make that the cut off point. Lastly, there are 1421 '0' values. Since eBay is a bidding site, we can remove those and atleast start from $1.

In [7]:
#filtering autos dataset
autos = autos[autos['price'].between(1, 350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [8]:
#exploring odometer_kg column
print(autos['odometer_km'].unique().shape)
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts().sort_index(ascending=False).head(30))

(13,)
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64


Based on the data, we can see that are only 13 unique values, so theres no need to sort the index as the data is small enough to observe. There are no obvious outliers in this data and while there are a lot more higher-valued numbers, there are enough smaller numbers to keep the data we have.

In [9]:
#caculating distribution of date_crawled
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We can see that the overall data is fairly spread amongst the dates, creating uniformity.

In [10]:
#calculating distribution of ad_created
(autos['ad_created']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
 .tail(40)
)

2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001483
2016-03-05    0.022897
2016-03-06    0.015320
2016-03-07    0.034737
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, dtype: float64

The distribution in this section also aligns with the 'date_crawled' section as after the 5th of March, the value counts increase and almost match the dates from that column as well.

In [11]:
#calculating distribution of last_seen
(autos['last_seen']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

There is a huge discrepancy between the last three dates and the rest of the data. Lets take a look at the registration_year to see if there could be anything that could've affected this jump.

In [12]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

So after taking a look at our registration_year data, we can see there the max and min do not make sense with the year of these car registrations. This data is clearly affecting our earlier observations - the spike in the last three years.

In [13]:
autos['registration_year'].value_counts().sort_index().tail(60)


1970      38
1971      26
1972      33
1973      23
1974      24
1975      18
1976      21
1977      22
1978      44
1979      34
1980      85
1981      28
1982      41
1983      51
1984      51
1985      95
1986      72
1987      72
1988     135
1989     174
1990     347
1991     339
1992     370
1993     425
1994     629
1995    1227
1996    1373
1997    1951
1998    2363
1999    2897
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

From observing the value count, we can see that most of our data lies within 1994-2016, so these values will become our boundary points.

In [14]:
autos = autos[autos['registration_year'].between(1994,2016)]

In [15]:
autos['registration_year'].describe()

count    43903.000000
mean      2003.993987
std          5.413965
min       1994.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [16]:
(autos['brand'].value_counts(normalize=True) > .05)

volkswagen         True
bmw                True
opel               True
mercedes_benz      True
audi               True
ford               True
renault           False
peugeot           False
fiat              False
seat              False
skoda             False
mazda             False
nissan            False
smart             False
citroen           False
toyota            False
hyundai           False
mini              False
volvo             False
mitsubishi        False
honda             False
kia               False
sonstige_autos    False
alfa_romeo        False
suzuki            False
porsche           False
chevrolet         False
chrysler          False
dacia             False
daihatsu          False
jeep              False
land_rover        False
subaru            False
daewoo            False
saab              False
jaguar            False
rover             False
lancia            False
lada              False
trabant           False
Name: brand, dtype: bool

After observing the value counts, we will limit our aggregations by those with a frequency higher than 5%.
Thus, we have 6 car brands to observe.

In [17]:
#storing unique values
top_6_brands = autos['brand'].value_counts(normalize = True)
most_common_brands = top_6_brands[top_6_brands > 0.05].index

#creating brand/mileage mean dictionary
brand_mean = {}
mileage_mean = {}

for b in most_common_brands:
    selected_rows = autos[autos['brand'] == b]
    avg_price = selected_rows['price'].mean()
    avg_mileage = selected_rows['odometer_km'].mean()
    brand_mean[b] = avg_price
    mileage_mean[b] = avg_mileage

brand_mean

{'audi': 9727.60005217845,
 'bmw': 8542.713554463555,
 'ford': 3478.0246006389775,
 'mercedes_benz': 8832.383325111,
 'opel': 2983.578469787585,
 'volkswagen': 5599.258060954064}

Based on our data, we can see that audi, on average, is the most expensive car brand out of the top 6 most common brands available in the dataset.

In [18]:
#creating dataframe

bmp_series = pd.Series(brand_mean)
bmm_series = pd.Series(mileage_mean)
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df['mileage_mean'] = bmm_series

df

mean_price   mileage_mean
audi           9727.600052  128357.683277
bmw            8542.713554  132399.672400
ford           3478.024601  125386.581470
mercedes_benz  8832.383325  130340.404539
opel           2983.578470  129955.660961
volkswagen     5599.258061  128733.988516

We can see that, after creating our own dataframe, the mileage on the cars does not vary much from brand to brand. 